Add Python free path to unitary synthesis rust function

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit updates the UnitarySynthesis transpiler pass's rust internals to add a path for running the pass without Python. This is a requirement for using the pass from the C API. The complexity is that part of the pass's core functionality is still in Python, mainly two decomposers the XXDecomposer and the QSD function. These code paths are super commonly used, although the QSD function will be necessary to decompose 3+ qubit unitaries this is an ok limitation for the initial C API especially since development of a rust QSD is in progress this is only a temporary limitation.

The path through the pass without Python removes the explicit py token from the main entrypoint and also adds a new rust entrypoint distinct from the Python caller that has a new argument run_python_decomposers. If this argument is set to `false` and either the xx decomposer or qsd path are triggered the function will now either skip it or return an error if there is no alternative.


### Details and comments

Part of #14451

mtreinish committed:
Add Python free path to unitary synthesis rust function

This commit updates the UnitarySynthesis transpiler pass's rust
internals to add a path for running the pass without Python. This is a
requirement for using the pass from the C API. The complexity is that
part of the pass's core functionality is still in Python, mainly two
decomposers the XXDecomposer and the QSD function. These code paths are
super commonly used, although the QSD function will be necessary to
decompose 3+ qubit unitaries this is an ok limitation for the initial
C API especially since development of a rust QSD is in progress this is
only a temporary limitation.

The path through the pass without Python removes the explicit py token
from the main entrypoint and also adds a new rust entrypoint distinct
from the Python caller that has a new argument run_python_decomposers.
If this argument is set to `false` and either the xx decomposer or qsd
path are triggered the function will now either skip it or return an
error if there is no alternative.

Part of #14451

mtreinish committed:
Merge branch 'main' into python-free-unitary

mtreinish committed:
Merge remote-tracking branch 'origin/main' into python-free-unitary

ShellyGarion commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -631,11 +663,21 @@ fn get_2q_decomposers_from_target(
+                    | StandardGate::CRY
+                    | StandardGate::CRZ
+                    | StandardGate::CPhase
+            )
```
would it be better to define these gates once and for all (previously they were called PARAM_SET in line 77)? 
they repeat in line 471 for example.
we should note that all these gates are locally equivalent (up to 1-qubit gates to RZZ), and hence should use the TwoQubitControlledUDecomposer.

mtreinish replied:
I did it this way because using it as a single match arm gives the compiler a better chance the static array treats it like a `[u8; n]`. Rust doesn't have a syntax to store this in an object, but I can use a macro to deduplicate the lists I'll do that.

ShellyGarion commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -658,8 +700,25 @@ fn get_2q_decomposers_from_target(
+                        | StandardGate::ECR
+                        | StandardGate::DCX
+                        | StandardGate::ISwap
+                        | StandardGate::CH
```
would it be better to define these gates once and for all? some of them repeat in line 462 for example.
previously they were called GOODBYE_SET in line 76. 
we should note that these gates should use the TwoQubitBasisDecomposer as default.
why DCX and iSwap appear here? all other gates are locally equivalent to CX (up to 1-qubit gates), but DCX and iSwap need two CX gates?
perhaps other standard gates should be added here as well?

eliarbel replied:
Adding to Shelly's comment: `GOODBYE_SET` originally included only `cx`, `cz` and `ecr`. 
Also note that when we construct a decomposer using `get_2q_decomposer_from_basis`,  `cx`, `cz`, `iswap` and `ecr` are used as basis names.

mtreinish replied:
I just went through our standard 2q gates and tried running the `TwoQubitBasisDecomposer` on all the gates that didn't require a parameter and added the ones which worked and these were part of it. In the path here realistically it either runs on `TwoQubitBasisDecomposer` or `ControlledUDecomposer` (in the python path there is still the xx decomposer but I was thinking in a rust free mode) so if the decomposer worked my thinking was we should just use it because otherwise we will use it after computing that we can. That being said maybe I was wrong to preclude the XX decomposer here I can add that back in a python context.

I was less concerned with the `basis_gates` path realistically we should eliminate it and just use a target (we can build a virtual target in python with just the gates) but for this PR I can extend the list if you think it's important

mtreinish replied:
I deduplicated this in: https://github.com/Qiskit/qiskit/pull/14732/commits/c4f3bdcaa6169d06b646af3baffbff6bbbd9c151 (and refined it a bit in: https://github.com/Qiskit/qiskit/pull/14732/commits/f3808782632821a15a0cf868575bd4f8f3c7e425) which I think should address concerns on code duplication. 

I also double checked of the gates I added here, DCX and ISWAP won't work with the XXDecomposer at all, so they only can work with the TwoQubitBasisDecomposer. CHGate would work with it, but I don't think it's a major loss if we don't try the XXDecomposer in the case of a CHGate.

ShellyGarion replied:
CHGate is locally equivalent (up to 1-qubit gates) to CXGate, so it makes sense to use the TwoQubitBasisDecomposer for it, see [here](https://github.com/Qiskit/qiskit/blob/17577912c02124f175807ec5d115bda23e58b6c5/qiskit/circuit/library/standard_gates/h.py#L211).

ShellyGarion replied:
I like your current approach of having the two sets PARAM_SET and TWO_QUBIT_SET and avoid code duplication.
It seems that the gates are indeed correct according to the original version of the code that checks if the gates are super controlled, but I would still like to verify this with Lev who wrote this code.
https://github.com/Qiskit/qiskit/blob/8887f44362b7709a7a872c6794b64e8bff99dc8c/qiskit/synthesis/two_qubit/two_qubit_decompose.py#L881

ShellyGarion commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -691,16 +750,35 @@ fn get_2q_decomposers_from_target(
+                    | StandardGate::ECR
+                    | StandardGate::DCX
+                    | StandardGate::ISwap
+                    | StandardGate::CH
```
same question as above

ShellyGarion reviewed (commented):
I have some questions on the lists of standard gates used as defaults in step 1 and step 2.

eliarbel commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -199,8 +193,7 @@ fn apply_synth_sequence(
 #[pyo3(name = "run_main_loop", signature=(dag, qubit_indices, min_qubits, target, basis_gates, synth_gates, coupling_edges, approximation_degree=None, natural_direction=None, pulse_optimize=None))]
-pub fn run_unitary_synthesis(
-    py: Python,
+pub fn py_unitary_synthesis(
```
I can't comment on the doc comment above - typo: elegible -> eligible

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14732/commits/4078caacc65e942f6d1850480bcc4b4417226d5d

eliarbel commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -212,10 +205,34 @@ pub fn run_unitary_synthesis(
+    approximation_degree: Option<f64>,
+    natural_direction: Option<bool>,
+    pulse_optimize: Option<bool>,
+    run_python_decomposers: bool,
```
I was wondering about this flag: it's used to gracefully error out instead of panicking when a Python interpreter is not initialized. Is there a use-case for this on the Python side (i.e. when calling `run_unitary_synthesis` directly)? Otherwise for the C API side maybe we should just assume no Python-dependent execution at this stage and just remove this flag?

mtreinish replied:
The graceful handling was more just because I had the context to do it rather than panic in the C context. In the follow-on PR adding the C API I used the error message text. The flag is also used in the two qubit decomposer to skip the xx decomposer path which requires python. I used the flag to escape the 2q decomposer function rather than ever trying to instantiate the xx decomposer. This path is still otherwise executable from C/Rust because the target could have a gate like `RZX(pi/8)` which is expressible in a target but isn't valid for the `TwoQubitBasisDecomposer` and in Python the `XXDecomposer` would be used.

eliarbel commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -227,52 +244,57 @@ pub fn run_unitary_synthesis(
+            let new_node_op: OperationFromPython = Python::with_gil(|py| {
+                let raw_blocks: Vec<PyResult<Bound<PyAny>>> = py_instr
+                    .instruction
+                    .getattr(py, "blocks")?
```
While we're at it - maybe we can use `PyInstruction::blocks` here and get rid of one more Python call?

mtreinish replied:
I think we can now that you moved vars and exprs to rust. Prior to this we would have lost that context. I guess it's still an issue for `metadata` and `name` but I don't think that's super important for control flow blocks. (there's also the deprecated unit and duration attributes but they don't apply to control flow). That being said it doesn't actually remove the python call it just hides it in `PyInstruction::blocks()`.

eliarbel commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -389,26 +411,35 @@ pub fn run_unitary_synthesis(
             }
             // Run 3q+ synthesis
             _ => {
+                if !run_python_decomposers {
```
In case `run_python_decomposers` stays, can this be moved inside the else block of the next if statement to continue execution in case there are no basis gates and target?

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/14732/commits/38e9ba1d69051fe766a656d96717d2ed718ec055

eliarbel commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -691,16 +750,35 @@ fn get_2q_decomposers_from_target(
+            return Ok(Some(decomposers));
+        } else {
+            return Err(QiskitError::new_err(
+                "Target has no gates available on qubits to synthesize over without python",
```
```suggestion
                "Target has no gates available on qubits to synthesize over without Python",
```

eliarbel commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -992,28 +1074,25 @@ fn reversed_synth_su4_sequence(
 
 /// Apply synthesis for decomposers that return a DAG (XX).
-fn synth_su4_dag(
+fn synth_su4_xx_decomposer(
```
A documentation fix following the rename here: the doc comment of `reversed_synth_su4_dag` mentions the old name as the calling the function.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/14732/commits/b25a8416b93ad8fbd607c242908816b6f2360a02

eliarbel commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -1166,7 +1248,6 @@ fn synth_error(
 /// the decompostion will use the given `basis_gates` and the first valid decomposition
```
```suggestion
/// the decomposition will use the given `basis_gates` and the first valid decomposition
```

eliarbel reviewed (commented):
Overall this looks good. I wasn't sure about the changes in the basis gates which are used for stopping the decomposers search, so I consulted a bit with Shelly. Please see her comments about it. 
I also left a couple questions and a few minor documentation change suggestions.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into python-free-unitary

mtreinish committed:
Duplicated decomposer matcher for StandardGate objects

This commit adds a custom macro to provide the matcher for determining
if a standard gate is supported by each decomposer. This provides a
single source of truth for the matcher which may be used more than once.

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

mtreinish committed:
Fix typo

mtreinish committed:
Evaluate 3q python path after target check

mtreinish committed:
Simplify basis_gates path logic

mtreinish committed:
Fix docstring reference to old xxdecomposer synth function

eliarbel reviewed (dismissed):
Thanks Matthew for the changes. I don’t have any additional comments, this looks ready now. I’ll approve but leave it unmerged in case @raynelfss has anything to add.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into python-free-unitary

ShellyGarion reviewed (dismissed):
I like the fact that we have clearly defined two sets of standard gates that are supported by the two 2-qubit rust decomposers.
By looking at the original version of the 2-qubit synthesis code, the lists of gates look correct to me.

mtreinish committed:
Merge branch 'main' into python-free-unitary

mtreinish committed:
Fix lint
